src/main/java/nextstep/courses/domain/Session.java

```diff
@@ -0,0 +1,131 @@
+package nextstep.courses.domain;
+
+import nextstep.payments.domain.Payment;
+
+import java.time.LocalDateTime;
+import java.util.Objects;
+
+public class Session {
+    private Long id;
+
+    private SessionImage image;
+
+    private SessionType type;
+
+    private SessionState state;
+
+    private int studentCount;
+
+    private int maxStudentCount;
+
+    private Long price;
+
+    private LocalDateTime startedAt;
+
+    private LocalDateTime endedAt;
+
+    public Session(Long id, SessionImage image, SessionType type, SessionState state, int maxStudentCount,
+                   Long price, LocalDateTime startedAt, LocalDateTime endedAt) {
+        validateSession(type, maxStudentCount, price);
+
+        this.id = id;
+        this.image = image;
+        this.type = type;
+        this.state = state;
+        this.studentCount = 0;
+        this.maxStudentCount = maxStudentCount;
+        this.price = price;
+        this.startedAt = startedAt;
+        this.endedAt = endedAt;
+    }
+
+    public static Session ofFree(SessionImage image, SessionState state, int maxStudentCount, Long price, LocalDateTime startedAt, LocalDateTime endedAt) {
+        return new Session(
+                0L,
+                image,
+                SessionType.FREE,
+                state,
+                maxStudentCount,
+                price,
+                startedAt,
+                endedAt
+        );
+    }
+
+    public static Session ofPaid(SessionImage image, SessionState state, int maxStudentCount, Long price, LocalDateTime startedAt, LocalDateTime endedAt) {
+        return new Session(
+                0L,
+                image,
+                SessionType.PAID,
+                state,
+                maxStudentCount,
+                price,
+                startedAt,
+                endedAt
+        );
+    }
+
+    private void validateSession(SessionType type, int maxStudentCount, Long price) {
+        if (type == SessionType.PAID) {
+            validatePaidSession(maxStudentCount, price);
+        }
+        if (type == SessionType.FREE) {
+            validateFreeSession(maxStudentCount, price);
+        }
```

Comment:
SessionType 별로 검증을 다르게 구현해주셨네요.
그렇다면, 세션종류가 검증의 책임을 가진다고도 볼수 있겠네요 🧐
enum 에서 함수형 인터페이스 `BiConsumer<Integer, Long>` 를 활용해보는건 어떨까요?

```suggestion
        type.validate(maxStudentCount, price);
```

Reply:
너무 좋은 방법이네요 :)
구현하다 궁금한점이 생겼는데요, validatePaidSessionEnrollment 함수도 세션종류가 유료일 때만 검증하는 로직인데 이 부분도 SessionType으로 옮겨야 할까요? 검증 값들이 Session 필드 값이라 책임을 넘기면 안될 것 같아서요!

Reply:
그게 적절해보이네요 👍